xdsl/dialects/builtin.py

```diff
@@ -1854,11 +1834,77 @@ def tensor_from_list(
             | Sequence[IntegerAttr[IntegerType]]
             | Sequence[AnyFloatAttr]
         ),
-        data_type: IntegerType | IndexType | AnyFloat,
+        data_type: _DenseElementT,
         shape: Sequence[int],
-    ) -> DenseIntOrFPElementsAttr:
+    ) -> DenseIntOrFPElementsAttr[_DenseElementT]:
         t = TensorType(data_type, shape)
-        return DenseIntOrFPElementsAttr.from_list(t, data)
+        return DenseIntOrFPElementsAttr[_DenseElementT].from_list(t, data)
+
+    @staticmethod
+    def parse_with_type(
+        parser: AttrParser, type: AttributeInvT
+    ) -> TypedAttribute[AttributeInvT]:
+        assert (
+            isa(type, VectorType[AnyDenseElement])
+            or isa(type, TensorType[AnyDenseElement])
+            or isa(type, MemRefType[AnyDenseElement])
+        )
+
+        return cast(
+            TypedAttribute[AttributeInvT],
+            parser.parse_dense_int_or_fp_elements_attr(type),
+        )
+
+    @staticmethod
+    def _print_one_elem(val: Attribute, printer: Printer):
+        if isinstance(val, IntegerAttr):
+            printer.print_string(f"{val.value.data}")
+        elif isinstance(val, FloatAttr):
+            printer.print_float(cast(AnyFloatAttr, val))
+        else:
+            raise Exception(
+                "unexpected attribute type "
+                "in DenseIntOrFPElementsAttr: "
+                f"{type(val)}"
+            )
+
+    @staticmethod
+    def _print_dense_list(
+        array: Sequence[AnyIntegerAttr] | Sequence[AnyFloatAttr],
+        shape: Sequence[int],
+        printer: Printer,
+    ):
+        printer.print_string("[")
+        if len(shape) > 1:
+            k = len(array) // shape[0]
+            printer.print_list(
+                (array[i : i + k] for i in range(0, len(array), k)),
+                lambda subarray: DenseIntOrFPElementsAttr._print_dense_list(
+                    subarray, shape[1:], printer
+                ),
+            )
+        else:
+            printer.print_list(
+                array,
+                lambda val: DenseIntOrFPElementsAttr._print_one_elem(val, printer),
+            )
+        printer.print_string("]")
+
+    def print_without_type(self, printer: Printer):
+        printer.print_string("dense<")
+        data = self.data.data
+        shape = self.get_shape() if self.shape_is_complete else (len(data),)
+        assert shape is not None, "If shape is complete, then it cannot be None"
+        if len(data) == 0:
+            pass
+        elif data.count(data[0]) == len(data):
+            DenseIntOrFPElementsAttr._print_one_elem(data[0], printer)
+        else:
+            DenseIntOrFPElementsAttr._print_dense_list(data, shape, printer)
+        printer.print_string(">")
+
+
+DenseIntElementsAttr: TypeAlias = DenseIntOrFPElementsAttr[IntegerType]
```

Comment:
This line feels like it could have its own PR, we could probably have a custom constraint to start with to minimise the diff

Reply:
I still feel like this could be its own PR :)

Reply:
You want a separate PR just for the type alias?

Reply:
Yeah that was the idea, it technically is its own API surface change and seems to be a large contribution to this PR's diff, independent of the other changes.

Reply:
Sorry that's what I mean, this change in the diff of this PR could already be done in main directly, and is not dependent on the changes you make here. Or am I misunderstanding?

Reply:
I don't see how to separate the changes. I can't change `DenseIntOrFPElementsAttr` to `DenseIntOrFPElementsAttr[AnyDenseElement]` before this PR.

I could potentially do two steps of changing every `DenseIntOrFPElementsAttr` to `DenseIntOrFPElementsAttr[AnyDenseElement]` and then do a second PR specialising some of these to `DenseIntElementsAttr` = `DenseIntOrFPElementsAttr[IntegerType]`, but I don't think the first diff will be any smaller.

Reply:
Ah I get it now, I could give that a go